File: pset6/readability/readability.py
from cs50 import get_string


def main():
    # Get user input text
    text = get_string("Text: ")
    
    # Calculation of Coleman_Liau index and relevant variables
    l = count_letters(text) / count_words(text) * 100
    s = count_sentences(text) / count_words(text) * 100
    coleman_liau_index = 0.0588 * l - 0.296 * s - 15.8
    
    # Printing grade levels
    if coleman_liau_index < 1:
        print("Before Grade 1")
    elif coleman_liau_index >= 16:
        print("Grade 16+")
    else:
        print("Grade {}".format(round(coleman_liau_index)))


# Count words by assuming space between words
def count_words(text):
    successive_letter_count = 0
    word_count = 0
    for char in text:
        if char.isalpha():
            successive_letter_count += 1
        # Checking that word consists of at least 1 successive alphabet
        if char == " " and successive_letter_count > 0:
            word_count += 1
            successive_letter_count = 0
    return word_count + 1


# Count letters by counting only alphabet characters
def count_letters(text):
    letter_count = 0
    for char in text:
        if char.isalpha():
            letter_count += 1
    return letter_count


# Count sentences using end-of-sentence punctuation
def count_sentences(text):
    sentence_count = 0
    for char in text:
        if char == '.' or char == '?' or char == '!':
            sentence_count += 1
    return sentence_count


if __name__ == "__main__":
    main()

File: pset2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int letter(string text);
int word(string text);
int sentence(string text);

int main(void)
{
    //Prompt to obtain text from user
    string text = get_string("Text: ");

    //Count number of letters in text
    int letter_count = letter(text);

    //Count number of words in text
    int word_count = word(text);

    //Count number of sentences in text
    int sentence_count = sentence(text);

    //Calculating Coleman-Liau index
    float L = (float)letter_count / word_count * 100;
    float S = (float)sentence_count / word_count * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;
    
    //Printing response according to various grades
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 1 && index < 16)
    {
        printf("Grade %i\n", (int)round(index));
    }
    else
    {
        printf("Grade 16+\n");
    }
}

//Abstraction: Count number of letters in text
int letter(string text)
{
    int letter_count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (islower(text[i]) || isupper(text[i]))
        {
            letter_count++;
        }
    }
    return letter_count;
}

//Abstraction: Count number of words in text
int word(string text)
{
    int space_count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            space_count++;
        }
    }
    return space_count + 1;
}

//Abstraction: Count number of sentences in text
int sentence(string text)
{
    int sentence_count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        switch (text[i])
        {
            case '.':
            case '?':
            case '!':
                sentence_count++;
        }
    }
    return sentence_count;
}